predicators/envs/doors.py

```diff
@@ -729,3 +729,447 @@ def _get_open_door_target_value(mass: float, friction: float,
         # A made up complicated function.
         return np.tanh(target_rot) * (np.sin(mass) +
                                       np.cos(friction) * np.sqrt(mass))
+
+
+class DoorKnobsEnv(DoorsEnv):
+    """A 2D navigation environment with obstacles, rooms, and doors."""
+    # Types
+    _robot_type = Type("robot", ["x", "y"])
+    _door_type = Type("door", ["x", "y", "theta", "mass", "friction", "open"])
+    _knob_type = Type("knob", ["x", "y", "theta", "rot", "target_rot", "open"])
+    _room_type = Type("room", ["x", "y"])
+    _obstacle_type = Type("obstacle", ["x", "y", "width", "height", "theta"])
+    open_door_thresh: ClassVar[float] = 0.1
+
+    def __init__(self, use_gui: bool = True) -> None:
+        super().__init__(use_gui)
+
+        # Predicates
+        self._InRoom = Predicate("InRoom", [self._robot_type, self._room_type],
+                                 self._InRoom_holds)
+        self._InDoorway = Predicate("InDoorway",
+                                    [self._robot_type, self._door_type],
+                                    self._InDoorway_holds)
+        self._InMainRoom = Predicate("InMainRoom",
+                                     [self._robot_type, self._room_type],
+                                     self._InMainRoom_holds)
+        self._DoorInRoom = Predicate("DoorInRoom",
+                                     [self._door_type, self._room_type],
+                                     self._DoorInRoom_holds)
+        # This predicate is needed as a precondition for moving from one
+        # door to another door in the same room.
+        self._DoorsShareRoom = Predicate("DoorsShareRoom",
+                                         [self._door_type, self._door_type],
+                                         self._DoorsShareRoom_holds)
+        # Static objects (always exist no matter the settings).
+        self._robot = Object("robby", self._robot_type)
+        # Hyperparameters from CFG.
+        self._room_map_size = CFG.doors_room_map_size
+        self._min_obstacles_per_room = CFG.doors_min_obstacles_per_room
+        self._max_obstacles_per_room = CFG.doors_max_obstacles_per_room
+        self._min_room_exists_frac = CFG.doors_min_room_exists_frac
+        self._max_room_exists_frac = CFG.doors_max_room_exists_frac
+        # See note in _sample_initial_state_from_map().
+        self._task_id_count = itertools.count()
+        self._door_to_knob: Dict[Object, Object] = {}
+        self._open_door_target_value = CFG.doorknobs_target_value
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "doorknobs"
+
+    def simulate(self, state: State, action: Action) -> State:
```

Comment:
there's a bunch of code that's being shared between this method and the same method in the superclass. Could you refactor slightly to introduce a helper function that handles the mechanics of computing the next state so that the main difference is in the mechanics of the door?